extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -97,27 +77,43 @@ export class AmpInstallServiceWorker extends AMP.BaseElement {
             'source (%s) or canonical URL (%s) of the AMP-document.',
             origin, sourceUrl.origin, canonicalUrl.origin);
         this.iframeSrc_ = iframeSrc;
-        this.insertIframe_();
+        this.scheduleIframeLoad_();
       }
+      return;
     }
-    return Promise.resolve();
-  }
 
-  /** @override */
-  renderOutsideViewport() {
-    // We want the service worker to be installed wherever the element is
-    // located in the document.
-    return true;
+    if (parseUrl(win.location.href).origin == parseUrl(src).origin) {
+      this.loadPromise(this.win).then(() => {
+        return install(this.win, src);
+      });
+    } else {
+      this.user().error(TAG,
+          'Did not install ServiceWorker because it does not ' +
+          'match the current origin: ' + src);
+    }
   }
 
-  /** @override  */
-  getLayoutPriority() {
-    return LayoutPriority.BACKGROUND;
+  /** @private */
+  scheduleIframeLoad_() {
+    Services.viewerForDoc(this.getAmpDoc()).whenFirstVisible().then(() => {
+      // If the user is longer than 10 seconds on this page, load
+      // the external iframe to install the ServiceWorker. The wait is
+      // introduced to avoid installing SWs for content that the user
+      // only engaged with superficially.
+      Services.timerFor(this.win).delay(() => {
+        this.deferMutate(this.insertIframe_.bind(this));
```

Comment:
@jridgewell removed deferMutate, make sure you merge the changes from here into this revert: https://github.com/ampproject/amphtml/commit/c4ef8697f5084ecbeef85d47317fc3831a643010#diff-d7bcc789dc4ca07564072200b4451bb3